### PR TITLE
Update dimuon L3 filter, for 2017 online muon reconstruction

### PR DESCRIPTION
**Pull request by BenjaminRS**
This PR complements that of #19465 for the dimuon L3 filter counterpart. It allows to:
- handle muons reconstructed by the full (L3FromL2 + L3FromL1) iterL3 sequence, and perform the matching with the candidate from the previous step
- introduce the possibility of disabling the matching of the L3 to the previous candidate (L2 or L1)

### CODE REVIEW

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 1]
```diff
@@ -458,3 +453,102 @@ HLTMuonDimuonL3Filter::triggeredByLevel2(TrackRef const & staTrack,vector<RecoCh
   }
   return ok;
 }
+
+bool HLTMuonDimuonL3Filter::applyMuonSelection(const RecoChargedCandidateRef& cand, const BeamSpot& beamSpot) const{
+	// eta cut
+	if (fabs(cand->eta())>max_Eta_) return false;
```
Comment (fwyzard) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: can you  use `std::abs` instead ?

**Comment by Martin-Grunewald** [comment]
Presumably we need a PR for 92X as well - please make one!

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

File: HLTrigger/Muon/interface/HLTMuonDimuonL3Filter.h [hunk 2]
```diff
@@ -12,11 +12,18 @@
...
-#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/RecoCandidate/interface/RecoChargedCandidate.h"
+#include "DataFormats/RecoCandidate/interface/RecoChargedCandidateFwd.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/Math/interface/deltaR.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
```
Comment (davidlange6) [inline HLTrigger/Muon/interface/HLTMuonDimuonL3Filter.h]: HI @BenjaminRS - please move includes not directly used in the header into the c++ file - at least deltaR and IdealMagneticFieldRecord appear to be in this scenario.

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 3]
```diff
@@ -199,55 +206,63 @@ HLTMuonDimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSet
...
 	  // ToDo: there should be a better way than dR,dPt matching
-	  const reco::Track& globalTrack = *link.globalTrack();
-	  float dR2 = deltaR2(tk->eta(),tk->phi(),globalTrack.eta(),globalTrack.phi());
-	  float dPt = std::abs(tk->pt() - globalTrack.pt())/tk->pt();
-          const TrackRef staTrack = link.standAloneTrack();
-	  if (dR2 < 0.02*0.02 and dPt < 0.001 and previousCandIsL2_) {
+	  const reco::Track& trackerTrack = *link.trackerTrack();
+
+	  float dR2 = deltaR2(tk->eta(),tk->phi(),trackerTrack.eta(),trackerTrack.phi());
+	  float dPt = std::abs(tk->pt() - trackerTrack.pt())/tk->pt();
```
Comment (davidlange6) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: careful about pt=0 tracks here. (was already an issue in the old code I realize.)

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 4]
```diff
@@ -403,20 +322,12 @@ HLTMuonDimuonL3Filter::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSet
 	      for (auto & i : vref) {
 		RecoChargedCandidateRef candref =  RecoChargedCandidateRef(i);
 		TrackRef tktmp = candref->get<TrackRef>();
-		if (tktmp==tk1) {
-		  i1done = true;
-		} else if (tktmp==tk2) {
-		  i2done = true;
-		}
+		if (tktmp==tk1) i1done = true;
+		else if (tktmp==tk2) i2done = true;	//why is this an elif?
```
Comment (davidlange6) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: is the comment yet to be resolved?

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc [hunk 5]
```diff
@@ -458,3 +453,102 @@ HLTMuonDimuonL3Filter::triggeredByLevel2(TrackRef const & staTrack,vector<RecoCh
...
+	    if (ptLx2<min_PtMin_[iv]) return false;
+	    if (ptLx2>max_PtMin_[iv]) return false;
+	  } else {
+	    if (ptLx2<min_PtMax_[iv]) return false;
+	    if (ptLx1<min_PtMin_[iv]) return false;
+	    if (ptLx1>max_PtMin_[iv]) return false;
+	  }
+	  if (pt12<min_PtPair_[iv]) return false;
+	  if (pt12>max_PtPair_[iv]) return false;
+	  proceed=true;
```
Comment (davidlange6) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: you can break out of this loop once proceed=true

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 4 above)
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: Hi David, it was more of a comment to myself - I was unsure when editing the code why it was written this way in the [old code](https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc#L408); I couldn't see how tktmp==tk1 and tktmp==tk2 could be true at the same time (therefore wouldnt need the else if). However as it is harmless, I left it that way

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 3 above)
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: added protection for this case

File: HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc (hunk 5 above)
Comment (BenjaminRS) [inline HLTrigger/Muon/src/HLTMuonDimuonL3Filter.cc]: added break into the loop

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**